### ISSUE TITLE
Exposing ephemeral storage metrics to prometheus 

### ISSUE THREAD

**Issue by NickrenREN** [kubernetes/kubernetes#69507]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

Now, we only expose PVC related volumes metrics to protheus, https://github.com/kubernetes/kubernetes/pull/51553
but we need ephemeral storage metrics too, such as EmptyDir usage and capacity info.

/assign 
/sig storage
/cc @jingxu97 @gnufied

**Comment by D-PENG** [comment]
> E1008 06:51:25.170819       1 manager.go:102] unable to fully collect metrics: [unable to fully scrape metrics from source kubelet_summary:kube-master: unable to fetch metrics from Kubelet kube-master (kube-master): Get http://kube-master:10255/stats/summary/: dial tcp: lookup kube-master on 10.254.0.2:53: no such host, unable to fully scrape metrics from source kubelet_summary:kube-backup: unable to fetch metrics from Kubelet kube-backup (kube-backup): Get http://kube-backup:10255/stats/summary/: dial tcp: lookup kube-backup on 10.254.0.2:53: no such host, unable to fully scrape metrics from source kubelet_summary:kube-worker: unable to fetch metrics from Kubelet kube-worker (kube-worker): Get http://kube-worker:10255/stats/summary/: dial tcp: lookup kube-worker on 10.254.0.2:53: no such host]

**Comment by anmolsachan** [comment]
@NickrenREN Is there any timeline for this or any release version targeted ?

**Comment by NickrenREN** [comment]
@anmolsachan sorry for the delay, i'll send a PR for this. But not sure if it can get in

**Comment by gnufied** [comment]
@NickrenREN is this a good idea? Don't you think exposing storage metrics for ephemeral storage via prometheus will cause explosion of metrics that aren't necessarily persistent?

**Comment by julienlim** [comment]
Some of the ephemeral storage use cases that come to mind include the following:
* understand the ephemeral storage utilization for chargeback/showback and/or capacity planning purposes
* determine how much ephemeral storage is being consumed in order to plan migration to persistant storage

**Comment by granular-ryanbonham** [comment]
Being able to expose ephemeral storage metrics in Kubernetes would be useful in particular in clusters there you may have LimitRange enforcing limits on ephemeral storage and want to provide a way for service owners to be able to view there usage.  Would vote to see this implemented.

**Comment by NickrenREN** [comment]
Any comments  @gnufied @jingxu97  ?

**Comment by gnufied** [comment]
@NickrenREN my opinion on this has not changed. But lets see what other people think. Also we should get input from sig-instrumentation.

cc @brancz @sjenning 

/sig instrumentation

**Comment by brancz** [comment]
I can potentially see a cardinality issue as well, but I generally also think these metrics would quite nice to have.

**Comment by gnufied** [comment]
Okay backing up - as I understand, these metrics for ephemeral volumes could be useful in-case 2MB of allocated storage space for emptydir backed volumes are filled up and users aren't aware? What is the use case we are looking at?

**Comment by brancz** [comment]
Or for capacity planning purposes.

**Comment by tbarrella** [comment]
To me it seems like pod-level ephemeral storage metrics would be useful to help avoid underutilization or eviction.

It's natural to set ephemeral-storage requests and limits to reduce the chance of pod eviction and enforce pod-level quotas. Without metrics for total ephemeral storage usage per pod, it takes more work to track whether requests and limits are set appropriately.

Are there already other ways to quickly evaluate and adjust ephemeral-storage requests and limits?

**Comment by ragav0102** [comment]
Looking at ways to get these metrics, which will help plan ephemeral-storage requests and limits and avoid over/under provisioning. 
@tbarrella Were you able to find any workaround?

**Comment by jingxu97** [comment]
I think in PosStats:VolumeStats, you could get any type of volume usage information including ephemeral ones?

**Comment by ragav0102** [comment]
@jingxu97 Yeah, hoping that there are others like me, it would be great if these stats are exposed as metrics.

**Comment by saad-ali** [comment]
It sounds like the problem is that in order for the mechanism to set ephemeral-storage limits (see [docs](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#local-ephemeral-storage), currently beta) to be useful users need a way to observe the current usage? That makes sense.

> I think in PosStats:VolumeStats, you could get any type of volume usage information including ephemeral ones?

Does this work for folks?

**Comment by NickrenREN** [comment]
I agree with @ragav0102 , we have ephemeral storage info in `PodStats`, but we also need to expose them as metrics so that users can observe and use them more easily and directly just like what we have done for persistent volumes.

**Comment by saad-ali** [comment]
It certainly would be to nice to also have this information in Prometheus 

But, as @gnufied and @brancz mentioned cardinality of the metrics is a major concern.

The number of ephemeral volumes (e.g. secret volumes, etc.) that a normal Kubernetes cluster creates is far larger then the number of persistent volumes. And unlike persistent volumes, which are created once and (generally) used for a long time, new ephemeral volumes are constantly created and used for short periods of time.

Per the [Prometheus documentation ](https://prometheus.io/docs/practices/naming/#labels) on labels:

> CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

Any ideas on how we can make this work without hitting this issue?

**Comment by tbarrella** [comment]
While there are many more ephemeral storage volumes than persistent volumes, wouldn't the cardinality be similar to the number of pods, for which there are already metrics per container?

**Comment by brancz** [comment]
/remove-lifecycle stale

On a second thought, since these metrics would be on the kubelet, the cardinality on this endpoint would be practically limited, and people for whom the cardinality is too high to ingest into prometheus can just drop them at ingest time via metric relabeling.

**Comment by cofyc** [comment]
> It certainly would be to nice to also have this information in Prometheus
> 
> But, as @gnufied and @brancz mentioned cardinality of the metrics is a major concern.
> 
> The number of ephemeral volumes (e.g. secret volumes, etc.) that a normal Kubernetes cluster creates is far larger then the number of persistent volumes. And unlike persistent volumes, which are created once and (generally) used for a long time, new ephemeral volumes are constantly created and used for short periods of time.
> 
> Per the [Prometheus documentation ](https://prometheus.io/docs/practices/naming/#labels) on labels:
> 
> > CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.
> 
> Any ideas on how we can make this work without hitting this issue?

If we expose storage metrics for volumes which consume local disk storage only, the cardinality can be zero if no pods using the following volumes:

https://github.com/jingxu97/kubernetes/blob/b618411f1edb98afe807866fb3a607356d72ba24/pkg/kubelet/server/stats/volume_stat_calculator.go#L170-L176 

The secret volume is not treated as ephemeral storage (both in summary API and evict manager) as it's backed by a tmpfs. 

From my perspective, expose the following information is useful in production:

- stats of volumes which consume local disk storage
- stats of other ephemeral storage counted in eviction manager, e.g. container rootfs usage

I think we can ignore some inline volumes like secrets.

**Comment by EvanKrall** [comment]
Is there _any_ way for users to ask Kubernetes for the current ephemeral-storage usage of a pod that will match what triggers pod eviction?

**Comment by eddieparker** [comment]
Echoing what others have mentioned but adding a use case: i have a grafana graph which plots current usage of a resource (memory and cpu, presently) against their limits and requests.  It makes it trivial to diagnose why eviction happens.  That said, there is no way to similarly graph ephemeral storage usage.

My current workflow is to just guess at a number and binary search until it just starts blowing up and then add padding, which isn't very efficient. 

I'd welcome simply plumbing through whatever calculation the eviction api must already be doing to prometheus to enable this functionality.

**Comment by MartinKirchner** [comment]
I can totally agree to @eddieparker .
What I basically need is a way to get a warning before a pod is evicted to check what consumes so much space and how can I fix this.
Please implement some metric for it. Could just be the quotient of limit and current consumption. I don't need it detailled for every volume.

**Comment by albertocavalcante** [comment]
Is this being considered? Thanks

**Comment by petros-d** [comment]
FWIW I put together a simple exporter that takes the metrics from the kubelet API and exposes them in Prometheus format for scraping, in case it's useful for anyone else who needs to track this in the short term. 

https://github.com/petros-d/kubelet-stats-exporter

**Comment by boniek83** [comment]
> FWIW I put together a simple exporter that takes the metrics from the kubelet API and exposes them in Prometheus format for scraping, in case it's useful for anyone else who needs to track this in the short term.
> 
> https://github.com/petros-d/kubelet-stats-exporter

What is lacking here and in the data exposed by kubernetes is ephemeral storage usage per container. Better than nothing, not quite there yet.

**Comment by karlhungus** [comment]
Another datapoint: we also wanted these metrics and built something nearly identical to petros-d https://github.com/salesforce/kubelet-summary-exporter/ the only material difference being that it runs as a daemonset.

**Comment by jmcgrath207** [comment]
well this is awkward, I end up making something similar to @karlhungus as a good learning project but forgot to check back on this thread :sweat_smile:

FWIW, here is mine https://github.com/jmcgrath207/k8s-ephemeral-storage-metrics

**Comment by xing-yang** [comment]
Need help investigate this, start with per-pod/per-container metrics and how to deal with cardinality problem with help from sig-instrumentation.

**Comment by brancz** [comment]
I don't think there is anything we can do about the cardinality of these metrics, it's just reflecting the reality, which is the number of ephemeral volumes that exist. Besides, we already have metrics for non-ephemeral volumes, so it we've accepted that we have the number of volumes in number of series. I think these metrics are extremely valuable to troubleshoot evictions, which we're flying totally blind on today, so I think these metrics are well worth their price.

**Comment by jbg** [comment]
The cardinality problem is not Kubernetes' problem and seems like scope creep. K8s already exposes many metrics that have excess cardinality if misused; administrators have plenty of ways of dealing with this already in prometheus itself.

**Comment by dgrisonnet** [comment]
Cardinility really depends on how precise we need to be with these metrics, but knowing which pod this volume belongs to should be the bare minimum, so as @brancz mentioned it should be bounded to the number of ephemeral volumes.

> Besides, we already have metrics for non-ephemeral volumes, so it we've accepted that we have the number of volumes in number of series.

The scenario is a bit different here though. The number of ephemeral volumes can grow way faster than the number of non-ephemeral ones since they have a way shorter lifespan. When you would have one PV for most of the cluster life, you would have X amount of ephemeral volumes that would depend on the number of time the pod was restarted.

I won't argue on how valuable these metrics can be since we all agree on that, but I wouldn't expect all Kubernetes users to be willing to be that price. An example where it might be worth would be in CI clusters running short-living pods that are mounting secrets and configmaps.

The most convenient solution here in my opinion would be to expose these metrics on a different endpoint than the default kubelet one. The idea would be that if users want this data, they will be able to scrape it from this new endpoint at their own risk/cost and existing users wouldn't have to be worried about that.

**Comment by dgrisonnet** [comment]
> The cardinality problem is not Kubernetes' problem and seems like scope creep. K8s already exposes many metrics that have excess cardinality if misused; administrators have plenty of ways of dealing with this already in prometheus itself.

It is though. The majority of users are trusting us to not expose metrics to the default endpoints that they are already scraping that would blow up their monitoring system or the cost of it.
You also have to understand that although Kubernetes exposes metrics in Prometheus format by default, there are some users that are not using Prometheus as their monitoring backend so they might not have the same protection mechanisms in place.

For the record there is also a native way in Kubernetes to reduce the cardinally of metrics, but it is not currently being used to allow exposing unbounded metrics, although it might be in the future. https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2305-metrics-cardinality-enforcement

**Comment by brancz** [comment]
Yeah but trusting here means that we don’t put request IDs in metrics, not a perfectly reasonable and important metric that is used by the kubelet for evictions. I also don’t think the pod restart is any different, the cardinality is no different, there may be some churn but I would categorize it as negligible.

Another thing worth noting I think only ephemeral volumes that are writable are even in scope here. No read only volume can cause evictions, hence I don’t think they’re worth tracking. We’re essentially only talking about emptydir.

**Comment by dgrisonnet** [comment]
> Another thing worth noting I think only ephemeral volumes that are writable are even in scope here. No read only volume can cause evictions, hence I don’t think they’re worth tracking. We’re essentially only talking about emptydir.

If we only need to track emptydir, the amount of data we need to expose should be tremendously reduced. Churn should be the only problem with that scope but as you mentioned we already have that with other kubelet metrics, so that sounds reasonable to me :+1:

**Comment by farodin91** [comment]
I used https://github.com/jmcgrath207/k8s-ephemeral-storage-metrics as basis for my tool: https://github.com/GDATASoftwareAG/kubelet-stats-metrics

**Comment by gjkim42** [comment]
> The most convenient solution here in my opinion would be to expose these metrics on a different endpoint than the default kubelet one. The idea would be that if users want this data, they will be able to scrape it from this new endpoint at their own risk/cost and existing users wouldn't have to be worried about that.

That^ is an interesting idea.

If the main problem here is cardinality, we could let the kubelet be able to configure which metrics to expose or add more endpoints other than the default metrics endpoint (/metrics).

**Comment by brancz** [comment]
I don't find multiple endpoints practical. I think it's better if there is as few endpoints as possible (ideally single one, but I understand that ship has sailed), and the consumer decides to drop metrics using metric relabel configs.

**Comment by dgrisonnet** [comment]
Re-assigning since this hasn't been worked on for the last 5 years.

/unassign @NickrenREN 
/assign @machine424

**Comment by machine424** [comment]
I bet GitHub won’t mind now.

**Comment by machine424** [comment]
I have https://github.com/kubernetes/kubernetes/pull/121489 that exposes the `emptyDir`s usages, do you think it's relevant to expose their `SizeLimit`, when set, as well?

**Comment by jmcgrath207** [comment]
@machine424 

FWIW, I have created a controller for this problem that might help as a reference for your work. Engineers have been using this in their production clusters.

https://github.com/jmcgrath207/k8s-ephemeral-storage-metrics

**Comment by lozbrown** [comment]
> > do you think it's relevant to expose their `SizeLimit`, when set, as well?
> 
> I'd love to see sizeLimit as well! Helps with plotting nice graphs.

potentially the limit and the requests... this would allow you to use metrics to create a percentage used graph in something like grafana.

background, i started looking into this stuff after getting issues with my pods getting evicted for using all the ephemeral storage.

**Comment by sathieu** [comment]
https://github.com/kubernetes/kubernetes/pull/121489#issuecomment-2405646869:

> Coming back from the SIG meeting, the group was in favor of this change and approved the cardinality of these metrics. I am assigning `@dashpole` for instrumentation and kubelet review.

**Comment by dongjiang1989** [comment]
@rexagod  I am happy to work on this if no else has started with it.

**Comment by frittentheke** [comment]
> [@rexagod](https://github.com/rexagod) I am happy to work on this if no else has started with it.

That's awesome @dongjiang1989. Lack of transparency around the use of ephemeral storage

Am I right to assume you will be picking up the PR by @machine424 (see his last comment https://github.com/kubernetes/kubernetes/pull/121489#issuecomment-2569863137)

A quick glimpse let's me assume that one was more focused on just emptyDir?

https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage and also https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-emphemeralstorage-consumption document that

> 
> - emptyDir volumes, except tmpfs emptyDir volumes
> - directories holding node-level logs
> - writeable container layers
> 

are all considered ephemeral storage.

With the option to hard limit ephemeral storage since 1.25 (https://kubernetes.io/blog/2022/09/19/local-storage-capacity-isolation-ga/) and with `LocalStorageCapacityIsolationFSQuotaMonitoring` being in BETA now, it seems only natural to first track the current usable patterns via metrics.

**Comment by dongjiang1989** [comment]
Thanks @frittentheke  Thanks for summarizing the information again.

I learned the PR [#121489](https://github.com/kubernetes/kubernetes/pull/121489) by @machine424 and above information.

Is it possible to split multiple PRs for submission?
* tmpfs emptyDir volume 
* local-ephemeral-storage
* directories holding node-level logs  and  writeable container layers

**Comment by jkroepke** [comment]
It seems that the PR #121489 is closed due inactivity. What makes this request so complicated, if the data is availible via `/stats/summary`, but not as metric?

**Comment by zanarellidev** [comment]
Picking up @brancz's point that kubelet-local cardinality is bounded: the data for ephemeral volumes is already collected today, it's just discarded before becoming a metric.

`volumeStatsCollector.CollectWithStability` (`pkg/kubelet/metrics/collectors/volume_stats.go`) iterates every `podStat.VolumeStats`, which includes secrets/configmaps/emptyDir, then does `if pvcRef == nil { continue }` and drops it. `VolumeStats.Name` (the pod-spec volume name, not per-secret-key data) is already on that struct for every volume, PVC-backed or not.

Proposal: emit a second metric set for the non-PVC branch, labeled `namespace, pod, volume_name` instead of `namespace, persistentvolumeclaim`. Cardinality is bounded by (pods on this node) x (volumes per pod), the same shape container-level kubelet metrics already ship at, not by secret/configmap count or content. Gate it behind a flag so clusters that can't afford even that pay nothing by default.

Different angle from #121489 (closed for inactivity, added a dedicated emptyDir-only collector): extending the existing volume_stats collector covers every non-PVC volume type in one pass instead of just emptyDir.

Happy to put up a PR for this if the approach sounds right.

**Comment by frittentheke** [comment]
@zanarellidev awesome to see this move forward!

**Comment by zanarellidev** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/141061 implementing this.
